catalog/lease: allow historical queries on dropped descriptors

fqazi opened the pull request:
Previously, when attempting to lease historical descriptors that were dropped we would always produce the "descriptor is dropped" error, once all leases are released on the prior version. This is problematic because with locked timestamp leasing, we can end up doing historical queries on dropped descriptors. To address this, this patch allows the lease manager to query historical descriptors on dropped objects. Dropped descriptor leases however are still not handed out.

This patch also tolerates missing targets (indexes and constraints) for comments within descriptors for SHOW CREATE. This is expected since leased descriptors may not fully line up with comments read from storage.

Fixes: #159583
Fixes: #159672

Release note: None

fqazi committed:
catalog/lease: allow historical queries on dropped descriptors

Previously, when attempting to lease historical descriptors that were
dropped we would always produce the "descriptor is dropped" error, once
all leases are released on the prior version. This is problematic
because with locked timestamp leasing, we can end up doing historical
queries on dropped descriptors. To address this, this patch allows the
lease manager to query historical descriptors on dropped objects.
Dropped descriptor leases however are still not handed out.

Fixes: #159583

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/159729)
<!-- Reviewable:end -->

fqazi committed:
sql: tolerate missing comment references to descriptor

Previously, when using non-leased descriptors it was guaranteed the
comments table would be synchronized with the descriptor we are
currently using. With leased descriptors this is no longer true, so SHOW
CREATE TABLE should tolerate missing targets for comments.

Informs: #159583

Release note: None

rafiss reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @rafiss made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159729#-:-Ogn-KY4-IJMrb5tw0QI:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained.



<!-- Sent from Reviewable.io -->
